model/src/application/interaction/application_command/data/mod.rs

```diff
@@ -26,50 +34,177 @@ pub struct CommandData {
-        value: bool,
-    },
-    SubCommand {
-        name: String,
-        #[serde(default)]
-        options: Vec<CommandDataOption>,
-    },
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct CommandDataOption {
+    pub name: String,
+    pub value: CommandOptionValue,
 }
 
-impl CommandDataOption {
-    pub const fn kind(&self) -> &'static str {
-        match self {
-            CommandDataOption::String { .. } => "String",
-            CommandDataOption::Integer { .. } => "Integer",
-            CommandDataOption::Boolean { .. } => "Boolean",
-            CommandDataOption::SubCommand { .. } => "SubCommand",
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum CommandOptionValue {
+    String(String),
+    Integer(i64),
+    Boolean(bool),
+    User(UserId),
+    Channel(ChannelId),
+    Role(RoleId),
+    Mentionable(GenericId),
+    SubCommand(Vec<CommandDataOption>),
+    SubCommandGroup(Vec<CommandDataOption>),
+}
+
+#[derive(Debug, Deserialize)]
+struct CommandDataOptionRaw<'a> {
+    name: String,
+    #[serde(rename = "type")]
+    kind: CommandOptionType,
+    value: Option<CommandOptionValueRaw<'a>>,
+    #[serde(default)]
+    options: Option<Vec<CommandDataOption>>,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+enum CommandOptionValueRaw<'a> {
+    String(Cow<'a, str>),
+    Integer(i64),
+    Boolean(bool),
+}
+
+impl Serialize for CommandDataOption {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut state = serializer.serialize_struct("CommandDataOptionRaw", 3)?;
+
+        state.serialize_field("name", &self.name)?;
+
+        state.serialize_field("type", &self.value.kind())?;
+
+        match self.value {
+            CommandOptionValue::SubCommand(ref opts)
+            | CommandOptionValue::SubCommandGroup(ref opts) => {
+                state.serialize_field("options", &Some(opts))?
+            }
+            CommandOptionValue::String(ref value) => state
+                .serialize_field("value", &Some(CommandOptionValueRaw::String(value.into())))?,
+            CommandOptionValue::Integer(value) => {
+                state.serialize_field("value", &Some(CommandOptionValueRaw::Integer(value)))?
+            }
+            CommandOptionValue::Boolean(value) => {
+                state.serialize_field("value", &Some(CommandOptionValueRaw::Boolean(value)))?
+            }
+            CommandOptionValue::User(UserId(id))
+            | CommandOptionValue::Channel(ChannelId(id))
+            | CommandOptionValue::Role(RoleId(id))
+            | CommandOptionValue::Mentionable(GenericId(id)) => state.serialize_field(
+                "value",
+                &Some(CommandOptionValueRaw::String(id.to_string().into())),
+            )?,
         }
+        state.end()
     }
+}
+
+impl<'de> Deserialize<'de> for CommandDataOption {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
```

Comment:
i also noticed this a bit later
```suggestion
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
```

Reply:
Oh, sorry. I let rust-analyzer generate those for me...